src/preamble.js

```diff
@@ -419,21 +419,31 @@ function stringToAscii(str, outPtr) {
   return writeAsciiToMemory(str, outPtr, false);
 }
 
-// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
+// Given a pointer 'ptr' to a UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
+// If len is passed, ptr points to an optionally null-terminated string of up to 'len' bytes.  If len is
+// not passed, ptr must point to a null-terminated string.
 
 #if TEXTDECODER
 var UTF8Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf8') : undefined;
 #endif
-function UTF8ArrayToString(u8Array, idx) {
+function UTF8ArrayToString(u8Array, idx, len) {
```

Comment:
This kind of change to add a `len` has been requested a number of times, and there is a very important subtlety to consider to well-define the `len` argument: what is the semantic that should occur when `len` is specified, but the input UTF8 array contains a null byte in the middle before `len` bytes are reached? Should a) the decoding continue past that null byte (writing the [null character](https://developer.mozilla.org/ro/docs/Web/JavaScript/Reference/Global_Objects/String) in the JS string) in the middle, up to exactly `len` bytes? or should b) the decoding stop at the null byte, and discard the remaining bytes (treating `len` rather more as a `maxLen` parameter).

Currently the changes in this PR implement b), but there is a problem that `TextDecoder` API actually implements a), so the implementation in this form has a bug that strings longer than 16 chars would get the behavior a), but strings shorter than that would get behavior b). I.e. TextDecoder is defined to allow null characters inside while converting, check this example:

```js
var d = new TextDecoder('utf-8');
var b = new Uint8Array([0x61, 0x6c, 0, 0x62]);
console.log(d.decode(b));
```

prints `al�b` (`al<nul>b`) in the browser.

If we want to define this as b), then we should make sure never to route to TextDecoder when `len` is specified, and perhaps rename `len` to `maxLen`. I'm leaning towards the semantic b) since that is more useful from security perspective, i.e. one can pass `UTF8ArrayToString` to convert strings with a predetermined known maximum buffer length, without having to count where the exact position of zero will be.

Reply:
Fair enough - I've implemented this by modifying the while loop to always search until a nul character is found or maxLen is reached.  This + TextDecoder should be quicker that doing it ourselves.